Name: Yusuf İpek
Student Number: 2220356048

In [3]:
#import the modules
import numpy as np
import pandas as pd

<h5>Read the excel file with pandas and copy the original data frame</h5>

In [4]:
df = pd.read_csv("16P.csv", encoding = "ISO-8859-1")
originalDF = df.copy()
originalDF

Response Id  You regularly make new friends.  \
0                0                                0   
1                1                                0   
2                2                                0   
3                3                                0   
4                4                                0   
...            ...                              ...   
59994        59994                                0   
59995        59995                                0   
59996        59996                                0   
59997        59997                                0   
59998        59998                                0   

       You spend a lot of your free time exploring various random topics that pique your interest  \
0                                                      0                                            
1                                                      0                                            
2                                                      0                                            
3                                                     -1                                            
4                                                      0                                            
...                                                  ...                                            
59994                                                 -1                                            
59995                                                  0                                            
59996                                                  0                                            
59997                                                  0                                            
59998                                                  0                                            

       Seeing other people cry can easily make you feel like you want to cry too  \
0                                                      0                           
1                                                     -2                           
2                                                      2                           
3                                                      3                           
4                                                     -1                           
...                                                  ...                           
59994                                                  0                           
59995                                                 -2                           
59996                                                  1                           
59997                                                  1                           
59998                                                  2                           

       You often make a backup plan for a backup plan.  \
0                                                    0   
1                                                   -3   
2                                                    0   
3                                                   -1   
4                                                    0   
...                                                ...   
59994                                               -3   
59995                                                1   
59996                                                0   
59997                                               -1   
59998                                                0   

       You usually stay calm, even under a lot of pressure  \
0                                                      0     
1                                                     -1     
2                                                     -1     
3                                                      0     
4                                                      2     
...                                                  ..

<h5>Drop the response id parameter</h5>

In [5]:
df = df.drop(["Response Id"], axis= "columns")
df

You regularly make new friends.  \
0                                    0   
1                                    0   
2                                    0   
3                                    0   
4                                    0   
...                                ...   
59994                                0   
59995                                0   
59996                                0   
59997                                0   
59998                                0   

       You spend a lot of your free time exploring various random topics that pique your interest  \
0                                                      0                                            
1                                                      0                                            
2                                                      0                                            
3                                                     -1                                            
4                                                      0                                            
...                                                  ...                                            
59994                                                 -1                                            
59995                                                  0                                            
59996                                                  0                                            
59997                                                  0                                            
59998                                                  0                                            

       Seeing other people cry can easily make you feel like you want to cry too  \
0                                                      0                           
1                                                     -2                           
2                                                      2                           
3                                                      3                           
4                                                     -1                           
...                                                  ...                           
59994                                                  0                           
59995                                                 -2                           
59996                                                  1                           
59997                                                  1                           
59998                                                  2                           

       You often make a backup plan for a backup plan.  \
0                                                    0   
1                                                   -3   
2                                                    0   
3                                                   -1   
4                                                    0   
...                                                ...   
59994                                               -3   
59995                                                1   
59996                                                0   
59997                                               -1   
59998                                                0   

       You usually stay calm, even under a lot of pressure  \
0                                                      0     
1                                                     -1     
2                                                     -1     
3                                                      0     
4                                                      2     
...                                                  ...     
59994                                                 -2     
59995                                                  3     
59996                    

<h5>Convert to values in csv file to numpy array for calculations</h5>

In [6]:
allValues = df.values
allValues

array([[0, 0, 0, ..., 0, 0, 'ENFP'],
       [0, 0, -2, ..., -1, 3, 'ISFP'],
       [0, 0, 2, ..., 2, 1, 'INFJ'],
       ...,
       [0, 0, 1, ..., 0, -1, 'ISTP'],
       [0, 0, 1, ..., 1, 0, 'ISTJ'],
       [0, 0, 2, ..., 0, -1, 'INFJ']], dtype=object)

<h5>Convert to personality types to integers</h5>

In [7]:
PERSONALITIES = {
"ESTJ": 0,
"ENTJ": 1,
"ESFJ": 2,
"ENFJ": 3,
"ISTJ": 4,
"ISFJ": 5,
"INTJ": 6,
"INFJ": 7,
"ESTP": 8,
"ESFP": 9,
"ENTP": 10,
"ENFP": 11,
"ISTP": 12,
"ISFP": 13,
"INTP": 14,
"INFP": 15
}

print(allValues[-1][-1])
for person in PERSONALITIES:
    allValues[allValues == person] = PERSONALITIES[person]
print(allValues[-1][-1])

INFJ
7


<h4>Feature Normalization</h4> <br/> 

![Formula](https://i.ibb.co/C7BM3JW/image.png)

In [8]:
valuesWithoutPersonality = allValues[:, :-1] # drop the personalities
normalizatedValues = (valuesWithoutPersonality - np.min(valuesWithoutPersonality)) / (np.max(valuesWithoutPersonality) - np.min(valuesWithoutPersonality))
print(normalizatedValues)
normalizatedValues = np.column_stack((normalizatedValues, allValues[:, -1:])) # add the personality to end of the array
print(normalizatedValues)

[[0.5 0.5 0.5 ... 0.5 0.5 0.5]
 [0.5 0.5 0.16666666666666666 ... 0.3333333333333333 0.3333333333333333
  1.0]
 [0.5 0.5 0.8333333333333334 ... 0.6666666666666666 0.8333333333333334
  0.6666666666666666]
 ...
 [0.5 0.5 0.6666666666666666 ... 0.3333333333333333 0.5
  0.3333333333333333]
 [0.5 0.5 0.6666666666666666 ... 0.5 0.6666666666666666 0.5]
 [0.5 0.5 0.8333333333333334 ... 0.6666666666666666 0.5
  0.3333333333333333]]
[[0.5 0.5 0.5 ... 0.5 0.5 11]
 [0.5 0.5 0.16666666666666666 ... 0.3333333333333333 1.0 13]
 [0.5 0.5 0.8333333333333334 ... 0.8333333333333334 0.6666666666666666 7]
 ...
 [0.5 0.5 0.6666666666666666 ... 0.5 0.3333333333333333 12]
 [0.5 0.5 0.6666666666666666 ... 0.6666666666666666 0.5 4]
 [0.5 0.5 0.8333333333333334 ... 0.5 0.3333333333333333 7]]


Get the test and train values, shuffle the list for working with different data.

In [9]:
def getValues(isNormalize = True, kFold = 1):
    global allValues
    if(isNormalize): allValues = normalizatedValues # normalizated
    values = allValues[ (kFold - 1) * 12000 :12000 * kFold]
    beforeTest = allValues[:(kFold - 1) * 12000]
    afterTest = allValues[12000 * kFold:]
    knownValues = np.row_stack((beforeTest , afterTest))
    print(f"Test Shape: {values.shape}")
    print(f"Train Shape: {knownValues.shape}")
    return values, knownValues

Calculating differences, finding nearest neighbor for given test, train values for giving K-values of list

In [10]:
def findPersonalityWithKNN(values, knownValues, K):
    personalities = []
    personalitiesK = []
    true = 0
    false = 0
    b = np.array(knownValues[:,:-1], dtype=np.float32)
    for value in values:
        a = np.array([value[:-1]], dtype=np.float32)
        distances = np.linalg.norm(a-b, axis=1)
        distances = np.column_stack((distances, knownValues[:,-1:]))
        distances = sorted(distances, key=lambda entry: entry[0])
        for kv in K:
            firstKdist = distances[:kv]
            distancesCounts = {}
            for d in firstKdist:
                try:
                    distancesCounts[d[1]] += 1
                except:
                    distancesCounts[d[1]] = 1
            distancesCounts =  dict(sorted(distancesCounts.items(), key=lambda item: item[1], reverse=True))
            personality = list(distancesCounts.keys())[0]
            personalitiesK.append(personality)
        pk = personalitiesK.copy()
        personalities.append(pk)
        personalitiesK.clear()
    updatedValues = np.column_stack((values, personalities))
    print("Add the founded nearest personalities to array...")
    print(f"Updated array shape: {updatedValues.shape}")
    return updatedValues

<h5>Calculation of the accuracy, precision and recall values for given data.</h5>
<img src="https://i.ibb.co/938q35N/image.png" alt="accuracy" style="height:100px;"/> <img src="https://i.ibb.co/J5B4XnR/image.png" alt="precision" style="height:100px;"/> <img src="https://i.ibb.co/hH8jJ2q/image.png" alt="recall" style="height:100px;"/>

Look for the founded personalities with knn and original personalities.</br>
True negative value is equal to sum of other personalities true positive values.

In [11]:
def findAccuracyOfValues(updatedValues, K, isNormalize):
    performanceEvulation = {y: {i: { "TP" : 0, "FP" : 0, "FN" : 0, "TN" : 0} for i in range(16) } for y in range(len(K))}
    
    #find the true positive, flase negative and false pozitive counts
    for k in range(len(K)):
        for value in updatedValues:
            if(value[-(len(K) + 1)] == value[-(len(K) - k)]):
                performanceEvulation[k][value[-(len(K) + 1)]]["TP"] += 1
            else:
                performanceEvulation[k][value[-(len(K) + 1)]]["FN"] += 1
                performanceEvulation[k][value[-(len(K) - k)]]["FP"] += 1
    # calculate the true negative count with addition of all other values true positive counts
    for k in range(len(K)):
        for key in performanceEvulation[k]:
            for key2 in performanceEvulation[k]:
                if key == key2: continue
                performanceEvulation[k][key]["TN"] += performanceEvulation[k][key2]["TP"]
    accuracies = []
    precisions = []
    recalls = []
    macroAccuracies = []
    # find the accuracy, precision, recall values. calculate them in micro average and macro average. 
    for k in range(len(K)):
        print(f"K: {2 * (k) + 1} \t\t\t Is Normalize: {isNormalize}")
        print("Micro: ")
        for key, v in PERSONALITIES.items():
            TP =  performanceEvulation[k][v]["TP"] 
            TN =  performanceEvulation[k][v]["TN"]
            FP =  performanceEvulation[k][v]["FP"]
            FN =  performanceEvulation[k][v]["FN"]
            accuracy = (TP + TN) / (TP + TN + FP + FN)
            precision = TP / (TP + FP)
            recall = TP / (TP + FN)
            accuracies.append(accuracy)
            precisions.append(precision)
            recalls.append(recall)
            print(f"{key}-> Accuracy: % {accuracy * 100} \t Precision: % {precision * 100} \t Recall: % {recall * 100}")    

        macroAccuracies.append( "% {:.2f}".format(sum(accuracies) / len(accuracies) * 100))
        print("\nMacro: ")
        print(f"Accuracies: {macroAccuracies[-1]}")
        print("Precision: % {:.2f}".format(sum(precisions) / len(precisions) * 100))
        print("Recall: % {:.2f}\n".format(sum(recalls) / len(recalls) * 100))
        accuracies = []
        precisions = []
        recalls = []
    #print a list which has all macro accuracy values
    for i in range(len(K)):
        print(f"K={K[i]}-> Accuracy: {macroAccuracies[i]}")
    print("") # last line

<h4>Time and Accuracy Comparison</h4> (Different k values and normalization)

Define a function for calculation

In [12]:
def calculate(isNormalize, K):
    global updatedValues
    values, knownValues = getValues(isNormalize=isNormalize, kFold= 1)
    updatedValues = findPersonalityWithKNN(values, knownValues, K = K)
    findAccuracyOfValues(updatedValues, K, isNormalize=isNormalize)

In [13]:
calculate(isNormalize=True, K = [1])

Test Shape: (12000, 61)
Train Shape: (47999, 61)
Add the founded nearest personalities to array...
Updated array shape: (12000, 62)
K: 1 			 Is Normalize: True
Micro: 
ESTJ-> Accuracy: % 99.6776382762131 	 Precision: % 97.36842105263158 	 Recall: % 97.62532981530343
ENTJ-> Accuracy: % 99.7453310696095 	 Precision: % 98.10366624525916 	 Recall: % 98.10366624525916
ESFJ-> Accuracy: % 99.7961610327841 	 Precision: % 98.00266311584554 	 Recall: % 98.79194630872483
ENFJ-> Accuracy: % 99.72839925309795 	 Precision: % 97.02702702702702 	 Recall: % 98.62637362637363
ISTJ-> Accuracy: % 99.70301230377598 	 Precision: % 97.55102040816327 	 Recall: % 97.68392370572208
ISFJ-> Accuracy: % 99.71147318397827 	 Precision: % 97.5609756097561 	 Recall: % 98.06451612903226
INTJ-> Accuracy: % 99.81311586816174 	 Precision: % 99.20318725099602 	 Recall: % 97.90301441677587
INFJ-> Accuracy: % 99.75379913405213 	 Precision: % 97.44245524296676 	 Recall: % 98.83268482490273
ESTP-> Accuracy: % 99.7961610327841 

In [14]:
calculate(isNormalize=False, K = [1])

Test Shape: (12000, 61)
Train Shape: (47999, 61)
Add the founded nearest personalities to array...
Updated array shape: (12000, 62)
K: 1 			 Is Normalize: False
Micro: 
ESTJ-> Accuracy: % 99.6776382762131 	 Precision: % 97.36842105263158 	 Recall: % 97.62532981530343
ENTJ-> Accuracy: % 99.7453310696095 	 Precision: % 98.10366624525916 	 Recall: % 98.10366624525916
ESFJ-> Accuracy: % 99.7961610327841 	 Precision: % 98.00266311584554 	 Recall: % 98.79194630872483
ENFJ-> Accuracy: % 99.72839925309795 	 Precision: % 97.02702702702702 	 Recall: % 98.62637362637363
ISTJ-> Accuracy: % 99.70301230377598 	 Precision: % 97.55102040816327 	 Recall: % 97.68392370572208
ISFJ-> Accuracy: % 99.71147318397827 	 Precision: % 97.5609756097561 	 Recall: % 98.06451612903226
INTJ-> Accuracy: % 99.81311586816174 	 Precision: % 99.20318725099602 	 Recall: % 97.90301441677587
INFJ-> Accuracy: % 99.75379913405213 	 Precision: % 97.44245524296676 	 Recall: % 98.83268482490273
ESTP-> Accuracy: % 99.7961610327841

In [15]:
calculate(isNormalize=True, K = [1,3,5])

Test Shape: (12000, 61)
Train Shape: (47999, 61)
Add the founded nearest personalities to array...
Updated array shape: (12000, 64)
K: 1 			 Is Normalize: True
Micro: 
ESTJ-> Accuracy: % 99.6776382762131 	 Precision: % 97.36842105263158 	 Recall: % 97.62532981530343
ENTJ-> Accuracy: % 99.7453310696095 	 Precision: % 98.10366624525916 	 Recall: % 98.10366624525916
ESFJ-> Accuracy: % 99.7961610327841 	 Precision: % 98.00266311584554 	 Recall: % 98.79194630872483
ENFJ-> Accuracy: % 99.72839925309795 	 Precision: % 97.02702702702702 	 Recall: % 98.62637362637363
ISTJ-> Accuracy: % 99.70301230377598 	 Precision: % 97.55102040816327 	 Recall: % 97.68392370572208
ISFJ-> Accuracy: % 99.71147318397827 	 Precision: % 97.5609756097561 	 Recall: % 98.06451612903226
INTJ-> Accuracy: % 99.81311586816174 	 Precision: % 99.20318725099602 	 Recall: % 97.90301441677587
INFJ-> Accuracy: % 99.75379913405213 	 Precision: % 97.44245524296676 	 Recall: % 98.83268482490273
ESTP-> Accuracy: % 99.7961610327841 

In [16]:
calculate(isNormalize=False, K = [1,3,5])

Test Shape: (12000, 61)
Train Shape: (47999, 61)
Add the founded nearest personalities to array...
Updated array shape: (12000, 64)
K: 1 			 Is Normalize: False
Micro: 
ESTJ-> Accuracy: % 99.6776382762131 	 Precision: % 97.36842105263158 	 Recall: % 97.62532981530343
ENTJ-> Accuracy: % 99.7453310696095 	 Precision: % 98.10366624525916 	 Recall: % 98.10366624525916
ESFJ-> Accuracy: % 99.7961610327841 	 Precision: % 98.00266311584554 	 Recall: % 98.79194630872483
ENFJ-> Accuracy: % 99.72839925309795 	 Precision: % 97.02702702702702 	 Recall: % 98.62637362637363
ISTJ-> Accuracy: % 99.70301230377598 	 Precision: % 97.55102040816327 	 Recall: % 97.68392370572208
ISFJ-> Accuracy: % 99.71147318397827 	 Precision: % 97.5609756097561 	 Recall: % 98.06451612903226
INTJ-> Accuracy: % 99.81311586816174 	 Precision: % 99.20318725099602 	 Recall: % 97.90301441677587
INFJ-> Accuracy: % 99.75379913405213 	 Precision: % 97.44245524296676 	 Recall: % 98.83268482490273
ESTP-> Accuracy: % 99.7961610327841

<h5>Time of calculations</h5>
Different k values and different normalizations </br>
<img src = "https://i.ibb.co/zxpsKcQ/image.png">
<img src = "https://i.ibb.co/YLjZ9fm/image.png">
</br>
<img src = "https://i.ibb.co/LxN9cr1/image.png">
<img src = "https://i.ibb.co/PgjBVyc/image.png">

<h5>Comparison</h5>
As you can see in outputs:<br/>
<ul>
  <li>For K = 1:
    <ul>
      <li>Both normalizated or not normalizated dataset the accuracy is same: %99.73</li>
      <li>For time: Normalizated dataset time is less than not normalizated data set (0.7 seconds)</li>
    </ul>
  </li>
  <li>
    For K = 1, 3, 5:
      <ul>
      <li>We can see that when the k value increases the accuracy increases too.</li>
      <li>For K = 1 Accuracy: %99.73</li>
      <li>For K = 3 Accuracy: %99.86</li>
      <li>For K = 5 Accuracy: %99.87</li>
      </ul>
  </li>
  <li>
  For time: I find the distances first after that, i look for k-nearest neighbors so there is not a big time changing in this calculation.
  </li>
</ul> 

<h4>Calculation For All Values</h4>
(K-fold = 1,2,3,4,5 ; k = 1,3,5,7,9, all values for normalizated data and not normalizated data)

In [17]:
K = [1,3,5,7,9]
kFold = [1,2,3,4,5]
for kf in kFold: # for splitting data
    for i in range(2): # for normalization
        print(f"K Fold: {kf}")
        print(f"Is Normalizations: {bool(i)}")
        values, knownValues = getValues(isNormalize=bool(i), kFold=kf)
        updatedValues = findPersonalityWithKNN(values,knownValues, K)
        findAccuracyOfValues(updatedValues, K, isNormalize=bool(i))


K Fold: 1
Is Normalizations: False
Test Shape: (12000, 61)
Train Shape: (47999, 61)
Add the founded nearest personalities to array...
Updated array shape: (12000, 66)
K: 1 			 Is Normalize: False
Micro: 
ESTJ-> Accuracy: % 99.6776382762131 	 Precision: % 97.36842105263158 	 Recall: % 97.62532981530343
ENTJ-> Accuracy: % 99.7453310696095 	 Precision: % 98.10366624525916 	 Recall: % 98.10366624525916
ESFJ-> Accuracy: % 99.7961610327841 	 Precision: % 98.00266311584554 	 Recall: % 98.79194630872483
ENFJ-> Accuracy: % 99.72839925309795 	 Precision: % 97.02702702702702 	 Recall: % 98.62637362637363
ISTJ-> Accuracy: % 99.70301230377598 	 Precision: % 97.55102040816327 	 Recall: % 97.68392370572208
ISFJ-> Accuracy: % 99.71147318397827 	 Precision: % 97.5609756097561 	 Recall: % 98.06451612903226
INTJ-> Accuracy: % 99.81311586816174 	 Precision: % 99.20318725099602 	 Recall: % 97.90301441677587
INFJ-> Accuracy: % 99.75379913405213 	 Precision: % 97.44245524296676 	 Recall: % 98.83268482490273


<h5>Time of all values calculation screenshot</h5>
<img src= "https://i.ibb.co/YthWxDh/image.png">

<h4>Analysis</h4>
<h5>Accuracy Of All Values</h5>
<ul>
    <li>If we take different parts of the dataset (different fold) as test data we can find different accuracy values but the difference is less than %0.1 .</li>
    <li>Also we can say that when the k values increasing the accuracies increas too.</li>
    <li>When we look for normalizated dataset or not normalizated dataset the accuracy values are same.</li>
    <li>Normalization does not change our accuracy values. (in my code at least)</li>
</ul>
<h5>Wrong Answers</h5>
<ul>
    <li>In this project we can make wrong classifications. There is some reasons for this mistake:</li>
    <ul>
        <li>We have 16 different personalities and we are looking for k-neighbors if we find the each neighbors is different, then we have a probably wrong guess.</li>
        <li>I mean we have multiple personalities and we don't get all guesses true because of the dataset and our algorithm.</li>
        <li>If our train set have more than one neighbors which repeatiton is equal to each other and more then remaining k-neighbors when we sort reversly and take the first element of k-neighbors' our founded personality is depending on chance.</li>
        <li>Also in K values equal to 1 (i mean less enough) our program looks for nearest neighbor, it can be wrong of course. When we look outputs, we can see that the nearest neighbor's personality is not always the same with actual personality.</li>
    </ul>
</ul>
<h5>Note</h5>
<ul>
    <li>If we increase the value of K we get more correct guesses. However when ve look for K = 7 and K = 9 we get same correctness so the key is not increase the value, the key is getting the optimum value.</li>
    <img src= "https://imgyukle.com/f/2023/01/15/JFv9fS.png">   
    <p>(Diffrent K values, different accuracy values)</p>
    <li>Also you can see the original data frame (readed from csv) and updated data frame below.</li>
</ul>

--------------------------------

<h5>Display the original data frame and updated data frame </h5>

In [18]:
columns = list(originalDF.columns)[1:] + ["K=1", "K=3", "K=5","K=7", "K=9"]
df = pd.DataFrame(updatedValues, columns=columns)
display(df)

You regularly make new friends.  \
0                            0.666667   
1                            0.333333   
2                                 0.5   
3                                 0.5   
4                                 0.5   
...                               ...   
11994                             0.5   
11995                             0.5   
11996                             0.5   
11997                             0.5   
11998                             0.5   

      You spend a lot of your free time exploring various random topics that pique your interest  \
0                                                    0.5                                           
1                                                    0.5                                           
2                                               0.666667                                           
3                                                    0.5                                           
4                                                    0.5                                           
...                                                  ...                                           
11994                                           0.333333                                           
11995                                                0.5                                           
11996                                                0.5                                           
11997                                                0.5                                           
11998                                                0.5                                           

      Seeing other people cry can easily make you feel like you want to cry too  \
0                                                    1.0                          
1                                               0.666667                          
2                                               0.333333                          
3                                               0.333333                          
4                                               0.333333                          
...                                                  ...                          
11994                                                0.5                          
11995                                           0.166667                          
11996                                           0.666667                          
11997                                           0.666667                          
11998                                           0.833333                          

      You often make a backup plan for a backup plan.  \
0                                            0.166667   
1                                            0.833333   
2                                            0.166667   
3                                                 0.0   
4                                            0.333333   
...                                               ...   
11994                                             0.0   
11995                                        0.666667   
11996                                             0.5   
11997                                        0.333333   
11998                                             0.5   

      You usually stay calm, even under a lot of pressure  \
0                                                    0.5    
1                                               0.333333    
2                                               0.166667    
3                                                    0.5    
4                                               0.833333    
...                                                  ...    
11994                                           0.166667    
11995                                                1.0    
11996                                           0.166667    
11997               

In [19]:
display(originalDF)

Response Id  You regularly make new friends.  \
0                0                                0   
1                1                                0   
2                2                                0   
3                3                                0   
4                4                                0   
...            ...                              ...   
59994        59994                                0   
59995        59995                                0   
59996        59996                                0   
59997        59997                                0   
59998        59998                                0   

       You spend a lot of your free time exploring various random topics that pique your interest  \
0                                                      0                                            
1                                                      0                                            
2                                                      0                                            
3                                                     -1                                            
4                                                      0                                            
...                                                  ...                                            
59994                                                 -1                                            
59995                                                  0                                            
59996                                                  0                                            
59997                                                  0                                            
59998                                                  0                                            

       Seeing other people cry can easily make you feel like you want to cry too  \
0                                                      0                           
1                                                     -2                           
2                                                      2                           
3                                                      3                           
4                                                     -1                           
...                                                  ...                           
59994                                                  0                           
59995                                                 -2                           
59996                                                  1                           
59997                                                  1                           
59998                                                  2                           

       You often make a backup plan for a backup plan.  \
0                                                    0   
1                                                   -3   
2                                                    0   
3                                                   -1   
4                                                    0   
...                                                ...   
59994                                               -3   
59995                                                1   
59996                                                0   
59997                                               -1   
59998                                                0   

       You usually stay calm, even under a lot of pressure  \
0                                                      0     
1                                                     -1     
2                                                     -1     
3                                                      0     
4                                                      2     
...                                                  ..